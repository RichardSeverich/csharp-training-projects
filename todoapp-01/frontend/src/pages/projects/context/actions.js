import { ActionTypes } from '../enum';
import noop from 'helpers/noop';
import { deleteData, getDataWithToken, postDataWithToken, updateData } from 'api/api';
import { idRoot } from 'helpers/constants';
import { getToken } from 'helpers/helper';

/**
 * @param {Function} dispatch function.
 * @param {object} payload new project.
 */
function OnProjectAdd(dispatch, payload) {
	postDataWithToken('projects', payload, getToken())
		.then(() => {
			OnProjectLoad(dispatch);
		})
		.catch((error) => {
			dispatch({ type: ActionTypes.ProjectError, payload: true });
			console.error(error);
		});
}

/**
 * @param {Function} dispatch function.
 * @param {object} payload project changed.
 */
function OnProjectChange(dispatch, payload) {
	updateData('projects/' + payload.uuid, payload.data, getToken())
		.then(() => {
			OnProjectLoad(dispatch);
		})
		.catch((error) => {
			dispatch({ type: ActionTypes.ProjectError, payload: true });
			console.error(error);
		});
}

/**
 * @param {Function} dispatch function.
 * @param {string} payload project uuid.
 */
function OnProjectDelete(dispatch, payload) {
	deleteData('projects/' + payload, getToken())
		.then(() => {
			OnProjectLoad(dispatch);
		})
		.catch((error) => {
			dispatch({ type: ActionTypes.ProjectError, payload: true });
			console.error(error);
		});
}

/**
 * @param {Function} dispatch function.
 */
export function OnProjectLoad(dispatch) {
	OnProjectLoading(dispatch, true);
	setTimeout(
		() =>
			getDataWithToken('projects', getToken())
				.then((projects) => {
					const initPath = [projects.find((p) => p.uuid === idRoot)];
					OnProjectSelect(dispatch, initPath);
					OnProjectLoading(dispatch, false);
					dispatch({ type: ActionTypes.ProjectLoad, payload: { projects } });
				})
				.catch((error) => {
					OnProjectLoading(dispatch, false);
					dispatch({ type: ActionTypes.ProjectError, payload: true });
					console.error(error);
				}),
		1000
	);
}

/**
 * @param {Function} dispatch function.
 * @param {boolean} payload is a state.
 */
function OnProjectLoading(dispatch, payload) {
	dispatch({ type: ActionTypes.ProjectLoading, payload });
}

/**
 * @param {Function} dispatch function.
 * @param {object[]} payload selected projects.
 */
function OnProjectSelect(dispatch, payload) {
	dispatch({ type: ActionTypes.ProjectSelect, payload });
}

/**
 * @param {Function} dispatch function.
 * @returns {object} callback functions.
 */
export default function ActionFactory(dispatch = noop) {
	return {
		onProjectAdd: (payload) => OnProjectAdd(dispatch, payload),
		onProjectChange: (payload) => OnProjectChange(dispatch, payload),
		onProjectDelete: (payload) => OnProjectDelete(dispatch, payload),
		onProjectLoad: () => OnProjectLoad(dispatch),
		onProjectLoading: (payload) => OnProjectLoading(dispatch, payload),
		onProjectSelect: (payload) => OnProjectSelect(dispatch, payload),
	};
}
