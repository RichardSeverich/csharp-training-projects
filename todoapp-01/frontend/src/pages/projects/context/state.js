import { ActionTypes } from '../enum';

const initState = {
	path: [
		{
			uuid: '00000000-0000-0000-0000-000000000000',
			name: 'My projects',
			parent: null,
			tasks: [],
		},
	],
	projects: [],
	tasks: [],
};

/**
 * @param {object} props props received.
 * @returns {object} returns state.
 */
export function State(props) {
	return { ...props, error: false, loading: false, data: initState };
}

/**
 * @param {object} state current state.
 * @param {object} action current action.
 * @returns {object} returns new state.
 */
export function Reducer(state, action) {
	const { data } = state;
	const { type, payload } = action;

	switch (type) {
		case ActionTypes.ProjectAdd:
			return { ...state, data: { ...payload } };
		case ActionTypes.ProjectChange:
			return { ...state, data: { ...payload } };
		case ActionTypes.ProjectDelete:
			return { ...state, data: { ...payload } };
		case ActionTypes.ProjectError:
			return { ...state, error: payload };
		case ActionTypes.ProjectLoad:
			return { ...state, data: { ...state.data, ...payload } };
		case ActionTypes.ProjectLoading:
			return { ...state, loading: payload };
		case ActionTypes.ProjectSelect:
			return { ...state, data: { ...data, path: payload } };
		default:
			return state;
	}
}
