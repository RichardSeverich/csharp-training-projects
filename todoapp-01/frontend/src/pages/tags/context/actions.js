import { ActionTypes } from '../enum';
import noop from 'helpers/noop';
import { getDataWithToken, postDataWithToken } from 'api/api';
import { getToken } from 'helpers/helper';

/**
 * @param {Function} dispatch function.
 * @param {string} payload is a tag name.
 */
function OnTagAdd(dispatch, payload) {
	const newTag = { name: payload };
	postDataWithToken('tags', newTag, getToken())
		.then(() => OnTagLoad(dispatch))
		.catch((error) => {
			console.error(error);
		});
}

/**
 * @param {Function} dispatch function.
 */
function OnTagLoad(dispatch) {
	OnTagLoading(dispatch, true);
	setTimeout(
		() =>
			getDataWithToken('tags', getToken())
				.then((tags) => {
					OnTagLoading(dispatch, false);
					dispatch({ type: ActionTypes.TagLoad, payload: { tags } });
				})
				.catch((error) => {
					OnTagLoading(dispatch, false);
					dispatch({ type: ActionTypes.TagError, payload: true });
					console.error(error);
				}),
		1000
	);
}

/**
 * @param {Function} dispatch function.
 * @param {boolean} payload is a state.
 */
function OnTagLoading(dispatch, payload) {
	dispatch({ type: ActionTypes.TagLoading, payload });
}

/**
 * @param {Function} dispatch function.
 * @param {object[]} payload selected tags.
 */
function OnTagSelect(dispatch, payload) {
	dispatch({ type: ActionTypes.TagSelect, payload });
}

/**
 * @param {Function} dispatch function.
 * @returns {object} callback functions.
 */
export default function ActionFactory(dispatch = noop) {
	return {
		OnTagAdd: (payload) => OnTagAdd(dispatch, payload),
		OnTagLoad: () => OnTagLoad(dispatch),
		OnTagLoading: (payload) => OnTagLoading(dispatch, payload),
		OnTagSelect: (payload) => OnTagSelect(dispatch, payload),
	};
}
