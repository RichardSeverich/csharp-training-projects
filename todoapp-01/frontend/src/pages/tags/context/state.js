import { ActionTypes } from '../enum';

const initState = {
	tags: [],
};

/**
 * @param {object} props props received.
 * @returns {object} returns state.
 */
export function State(props) {
	return { ...props, error: false, loading: false, data: initState };
}

/**
 * @param {object} state current state.
 * @param {object} action current action.
 * @returns {object} returns new state.
 */
export function Reducer(state, action) {
	const { data } = state;
	const { type, payload } = action;

	switch (type) {
		case ActionTypes.TagAdd:
			return { ...state, data: { ...payload } };
		case ActionTypes.TagError:
			return { ...state, error: payload };
		case ActionTypes.TagLoad:
			return { ...state, data: { ...state.data, ...payload } };
		case ActionTypes.TagLoading:
			return { ...state, loading: payload };
		case ActionTypes.TagSelect:
			return { ...state, data: { ...data, selected: payload } };
		default:
			return state;
	}
}
