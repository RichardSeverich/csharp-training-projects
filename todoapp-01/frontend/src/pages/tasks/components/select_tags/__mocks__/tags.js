let mockTags = [
	{ id: 1, name: 'Research', tasks: [] },
	{ id: 2, name: 'Today', tasks: [] },
	{ id: 3, name: 'Teach', tasks: [] },
	{ id: 4, name: 'Exercise', tasks: [] },
	{ id: 12, name: 'Tag', tasks: [] },
	{ id: 14, name: 'My tag', tasks: [] },
	{ id: 15, name: 'New tag', tasks: [] },
];

export default mockTags;
