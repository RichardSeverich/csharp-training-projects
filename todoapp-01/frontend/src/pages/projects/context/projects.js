import React, { createContext, useContext, useEffect, useMemo, useReducer } from 'react';
import { State, Reducer } from './state';
import ActionFactory from './actions';

export const ProjectContext = createContext();

export const ProjectDispatch = createContext();

/**
 * @param {object} props props received.
 * @returns {React.Component} returns react component.
 */
export function ProjectProvider(props) {
	const { children } = props;
	const [value, dispatch] = useReducer(Reducer, props, State);
	const actions = useMemo(() => ActionFactory(dispatch), [dispatch]);

	useEffect(() => {
		actions.onProjectLoad();
	}, [actions]);

	return (
		<ProjectContext.Provider value={value}>
			<ProjectDispatch.Provider value={actions}>{children}</ProjectDispatch.Provider>
		</ProjectContext.Provider>
	);
}

/**
 * @returns {object[]} returns state and actions.
 */
export function useProjectProvider() {
	const state = useContext(ProjectContext);
	const actions = useContext(ProjectDispatch);

	return [state, actions];
}
