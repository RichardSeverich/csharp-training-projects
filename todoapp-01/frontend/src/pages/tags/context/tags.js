import React, { createContext, useContext, useEffect, useMemo, useReducer } from 'react';
import { State, Reducer } from './state';
import ActionFactory from './actions';

export const TagContext = createContext();

export const TagDispatch = createContext();

/**
 * @param {object} props props received.
 * @returns {React.Component} returns react component.
 */
export function TagProvider(props) {
	const { children } = props;
	const [value, dispatch] = useReducer(Reducer, props, State);
	const actions = useMemo(() => ActionFactory(dispatch), [dispatch]);

	useEffect(() => {
		actions.OnTagLoad();
	}, [actions]);

	return (
		<TagContext.Provider value={value}>
			<TagDispatch.Provider value={actions}>{children}</TagDispatch.Provider>
		</TagContext.Provider>
	);
}

/**
 * @returns {object[]} returns state and actions.
 */
export function useTagProvider() {
	const state = useContext(TagContext);
	const actions = useContext(TagDispatch);

	return [state, actions];
}
