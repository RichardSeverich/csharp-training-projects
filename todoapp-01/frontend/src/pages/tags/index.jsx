import React, { useState, useEffect } from 'react';
import Header from 'components/header';
import MenuPages from 'components/menu';
import TagView from './views/main_view';
import { Form, Button } from 'semantic-ui-react';
import { Link } from 'react-router-dom';
import { useAppProvider } from 'context/application/app';
import { useTagProvider } from './context/tags';

const options = [
	{ key: 'asc n', text: 'Order by name (ascendent)', value: 'asc name' },
	{ key: 'desc n', text: 'Order by name (descendent)', value: 'desc name' },
	{ key: 'asc c', text: 'Order by number of tasks (ascendent)', value: 'asc count' },
	{ key: 'desc c', text: 'Order by number of tasks (descendent)', value: 'desc count' },
];
/**
 * @returns {object} Return page of tags available.
 */
function Tags(props) {
	const [state] = useAppProvider();
	const [, actions] = useTagProvider();
	const [show, setShow] = useState('asc name');

	useEffect(() => {
		actions.OnTagLoad();
	}, [state]);

	const handleChange = (e, { value }) => {
		setShow(value);
	};

	return (
		<div>
			<Header/>
			<MenuPages/>
				<h2>Tags available</h2>
				<Button to={'/tags/form'} as={Link}>Add Tag</Button>
				<Form>
					<Form.Select
						options={options}
						placeholder="Show or order..."
						name="show"
						value={show}
						onChange={handleChange}
					/>
				</Form>
				<TagView show={show}/>
		</div>
	);
}

export default Tags;
