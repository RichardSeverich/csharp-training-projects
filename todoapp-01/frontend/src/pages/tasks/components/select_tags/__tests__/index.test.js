import React from 'react';
import SelectTags from '../.';
import { render } from '@testing-library/react';
import { TagProvider } from 'pages/tags/context/tags';

describe('pages/tasks/components/select_tags', () => {
	describe('html render', () => {
		it('should render', () => {
			const { container } = render(
				<TagProvider>
					<SelectTags />
				</TagProvider>
			);
			expect(container.firstChild).toBeInstanceOf(HTMLDivElement);
		});

		it('should render properly', () => {
			const { container } = render(
				<TagProvider>
					<SelectTags />
				</TagProvider>
			);
			expect(container.firstChild).toMatchSnapshot();
		});
	});
});
