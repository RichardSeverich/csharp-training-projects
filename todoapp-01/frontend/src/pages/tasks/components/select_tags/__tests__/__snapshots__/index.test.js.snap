// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/tasks/components/select_tags html render should render properly 1`] = `
<div
  aria-expanded="false"
  class="ui fluid multiple search selection dropdown"
  id="select_tags"
  role="combobox"
>
  <input
    aria-autocomplete="list"
    autocomplete="off"
    class="search"
    tabindex="0"
    type="text"
    value=""
  />
  <span
    class="sizer"
  />
  <div
    aria-atomic="true"
    aria-live="polite"
    class="divider default text"
    role="alert"
  >
    Add tags or choose an existent tags
  </div>
  <i
    aria-hidden="true"
    class="dropdown icon"
  />
  <div
    aria-multiselectable="true"
    class="menu transition"
    role="listbox"
  >
    <div
      class="message"
    >
      No results found.
    </div>
  </div>
</div>
`;
